lib/spack/spack/test/concretize.py

```diff
@@ -190,6 +191,28 @@ def test_different_compilers_get_different_flags(self):
         assert set(client.compiler_flags['fflags']) == set(['-O0'])
         assert not set(cmake.compiler_flags['fflags'])
 
+    def test_architecture_inheritance(self):
+        spec = Spec('cmake-client %gcc@4.7.2 os=fe ^ cmake')
+        try:
+            spec.concretize()
+        except UnavailableCompilerVersionError:
+            # Concretization will fail with an UnavailableCompilerVersionError
+            # if the architecture is concretized incorrectly.
+            assert False
+        assert spec['cmake'].architecture == spec.architecture
+
+    def test_architecture_deep_inheritance(self):
+        # This test is dependent on traversal towards the root from mpich
```

Comment:
@becker33 IMO it's worth mentioning what this is testing, e.g.: "make sure that transitive dependencies receive root target when...".

> May only work because mpi is virtual, or because of artifacts of the ordering.

Is this behavior essential? it sounds like it is uncertain what is producing this effect, so IMO it may not be important to test it.

Reply:
@scheibelp I'm not entirely sure what is producing the effect, but I am sure that the previous behavior is wrong. In this example, clearly all of the dependencies should inherit the target from the root, and they didn't previously (both in the mainline and more complicatedly in my first commit of this PR). I added this as a regression for that behavior. I can add more detail about exactly what I think was going wrong previously if you think that's worthwhile, but it could become a very involved explanation.

Reply:
> I can add more detail about exactly what I think was going wrong previously if you think that's worthwhile, but it could become a very involved explanation.

To be clear I wasn't looking for an explanation of what went wrong previously. It sounded like it wasn't certain how this is now working. At least that's the impression I get from

> May only work because mpi is virtual, or because of artifacts of the ordering.

If there's an effect we *want* to happen. I think it should be deliberate and well understood how to make it happen. And the above comment implies to me that this test is succeeding but that it is not entirely understood how. Am I mistaken there?

Reply:
@scheibelp I'm sorry, my comment was unclear. I'm not sure why this test works as a test, not why it succeeds now. I'm unclear what it is about the test case that causes it to fail previously, so I can't be fully certain that it will fail on a regression if other aspects of Spack change. 

I may have confused matters by adding a comment at all. The short of the matter is that it works now, and unless something else changes, this test will catch a regression on that issue.

Reply:
@becker33 

> I'm not sure why this test works as a test

That confuses me.

A couple points though:

1. I think it's worth adding a comment explaining what this test is checking for (e.g. "ensure that all specs are concretized with the same target when..."
1. The comment as is implies that it isn't known how the test succeeds (e.g. do you know whether the ordering of the parent traversal is arbitrary from run to run?) so if it is in fact certain then IMO the comment should be removed.

Reply:
I'm confused by this discussion.  Does @becker33 know what the tests does or what "that issue" is?  If not, he should.  If so, it should go in a descriptive comment in a docstring so that the next guy knows what the test does.  The current comment here should really be cleaned up and made into a docstring.